pyproject.toml

```diff
@@ -3,7 +3,7 @@ requires = [
     "setuptools>=42",
     "setuptools_scm",
     "wheel",
-    "oldest-supported-numpy",
+    "numpy",
```

Comment:
We definitely don't want to change this.  We just implemented this so that this package will build and work at runtime with `numpy<1.20` under multiple versions of Python.  What is the reason for changing this?

This

Reply:
That makes sense, I wondered about that. Given that, I'm wondering how to make the package build succeed, since it fails otherwise?

Reply:
How is it failing?  Under all the platforms and python versions in the matrix?  Or only some?

It succeeds in building for all the tested configurations in the CI under linux as of 3 days ago.  See

https://github.com/spacetelescope/stsci.image/actions/runs/1277177708

And locally on my MacOS intel laptop I can easily build via `python -m build .` under py37-py39.

Reply:
When building against `oldest-supported-numpy` instead of `numpy`, I see [the following failure](https://github.com/zanecodes/stsci.image/runs/3745686887?check_suite_focus=true) while building the wheel:

```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DNUMPY=1 -I/tmp/pip-build-env-44omvv_1/overlay/lib/python3.6/site-packages/numpy/core/include -I/opt/python/cp36-cp36m/include/python3.6m -c src/_combinemodule.c -o build/temp.linux-x86_64-3.6/src/_combinemodule.o
    In file included from /tmp/pip-build-env-44omvv_1/overlay/lib/python3.6/site-packages/numpy/core/include/numpy/ndarraytypes.h:1809,
                     from /tmp/pip-build-env-44omvv_1/overlay/lib/python3.6/site-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                     from /tmp/pip-build-env-44omvv_1/overlay/lib/python3.6/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                     from src/_combinemodule.c:2:
    /tmp/pip-build-env-44omvv_1/overlay/lib/python3.6/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:15:2: warning: #warning "Using deprecated NumPy API, disable it by " "#defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
     #warning "Using deprecated NumPy API, disable it by " \
      ^~~~~~~
    In file included from /tmp/pip-build-env-44omvv_1/overlay/lib/python3.6/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                     from src/_combinemodule.c:2:
    src/_combinemodule.c: In function ‘_Py_combine’:
    src/_combinemodule.c:254:53: error: ‘NPY_ARRAY_INOUT_ARRAY2’ undeclared (first use in this function); did you mean ‘NPY_ARRAY_INOUT_ARRAY’?
                                                         NPY_ARRAY_INOUT_ARRAY2);
                                                         ^~~~~~~~~~~~~~~~~~~~~~
```

I'm not sure how else to resolve this though, I don't know enough about this package or its build process.

Reply:
Interesting.  I see that as a warning with `clang`:

```
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/jdavies/miniconda3/envs/test/include -arch x86_64 -I/Users/jdavies/miniconda3/envs/test/include -fPIC -O2 -isystem /Users/jdavies/miniconda3/envs/test/include -arch x86_64 -DNUMPY=1 -I/var/folders/jg/by5st33j7ps356dgb4kn8w900001n5/T/build-env-j_wb82nq/lib/python3.9/site-packages/numpy/core/include -I/var/folders/jg/by5st33j7ps356dgb4kn8w900001n5/T/build-env-j_wb82nq/include -I/Users/jdavies/miniconda3/envs/test/include/python3.9 -c src/_combinemodule.c -o build/temp.macosx-10.9-x86_64-3.9/src/_combinemodule.o
In file included from src/_combinemodule.c:2:
In file included from /var/folders/jg/by5st33j7ps356dgb4kn8w900001n5/T/build-env-j_wb82nq/lib/python3.9/site-packages/numpy/core/include/numpy/arrayobject.h:4:
In file included from /var/folders/jg/by5st33j7ps356dgb4kn8w900001n5/T/build-env-j_wb82nq/lib/python3.9/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
In file included from /var/folders/jg/by5st33j7ps356dgb4kn8w900001n5/T/build-env-j_wb82nq/lib/python3.9/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822:
/var/folders/jg/by5st33j7ps356dgb4kn8w900001n5/T/build-env-j_wb82nq/lib/python3.9/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: 
      "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION"
      [-W#warnings]
#warning "Using deprecated NumPy API, disable it with " \
 ^
src/_combinemodule.c:279:20: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
    for (i=0,f=0; i<(sizeof(functions)/sizeof(functions[0])); i++)
                  ~^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/_combinemodule.c:89:1: warning: unused function '_inner_old_median' [-Wunused-function]
_inner_old_median(int goodpix, int nlow, int nhigh, int ninputs,
^
3 warnings generated.
```

Reply:
Can we try an old version of ubuntu instead of the latest and see if we get the same error?  Try `ubuntu-18.04` perhaps?  I wondering if old gcc can handle this and newer cannot?  We should probably file an issue for this as well, so it can be fixed.

Reply:
That said, the current CI builds the C code without problem on `ubuntu-latest`.  So it is clearly possible.  What's the difference between the env setup between what you're doing and the current CI?

Reply:
Here's what our CI does.

https://github.com/spacetelescope/stsci.image/blob/62c38ae5c594df6ad0b4e969d0ec506b9a63b730/.github/workflows/ci.yml#L16-L42

It is very simple.

Reply:
py3.6 is generally ick and support ends in december, I see that referenced in one of zanes builds, and it feels like it shouldn't have been there.....

Reply:
> That said, the current CI builds the C code without problem on `ubuntu-latest`. So it is clearly possible. What's the difference between the env setup between what you're doing and the current CI?

The wheel build uses the version of `numpy` [specified in the `pyproject.toml`](https://github.com/spacetelescope/stsci.image/blob/master/pyproject.toml#L6), and the version of `numpy` specified by `oldest-supported-numpy` for the failing build [(on CPython 3.6)](https://github.com/zanecodes/stsci.image/runs/3745686887?check_suite_focus=true#step:2:600) as of the build date seems to [be `1.13.3`](https://github.com/scipy/oldest-supported-numpy/blob/v0.10/setup.cfg#L45), which is not in the CI build/test matrix, so this explains why the tests succeed; they also don't build against Python 3.6.

Should I exclude Python 3.6 from the wheel build then, since it seems to be unsupported by this package?

Reply:
> py3.6 is generally ick and support ends in december, I see that referenced in one of zanes builds, and it feels like it shouldn't have been there.....

Given that support for CPython 3.6 ends soon and this package does not support it anyhow, should we consider excluding CPython 3.6 from our wheel builds organization-wide, from [within our custom step](https://github.com/spacetelescope/action-publish_to_pypi/blob/master/build-wheel/action.yml#L31)?
Or would we prefer to exclude it only for this specific package, within [its workflow](https://github.com/spacetelescope/stsci.image/pull/23/files#diff-75e8bbdd253f4a71a14114808840e63b6a9c23529499687d148025d860336504R23)?

Reply:
Yeah, let's not build wheels for py36 here.  Your choice on the org-wide custom step.  But certainly a good number of our packages already don't support 3.6, i.e. the `jwst` calibration pipeline, `romancal`, etc.  The `sdist` will always be available for 3.6.

Reply:
But yeah, if we run into the wheel build problem here, we probably will elsewhere trying to build against 3.6.

Reply:
Here's a solution that should work.  Since this package has always depended on `numpy>=1.14` and since `oldest-supported-numpy` gives an older version for py36, how about the following on pyproject.toml:

```
    "oldest-supported-numpy; python_version>='3.7'",
    "numpy==1.14.6; python_version<'3.7'",
```

It builds under py36 for me with the above.  Hopefully it will for linux as well.

Feel free to add a py36 job to our CI matrix here to verify it works.

Reply:
Ok, well I went ahead and did a PR myself and fixed the py36 build for this package.  See

https://github.com/spacetelescope/stsci.image/pull/24

Thanks for the discussion and pointing out that `oldest-supported-numpy` was using numpy 1.13.x  👍

Reply:
I.e. feel free to rebase the branch for this PR against master, excising your changes to pyproject.toml.  Hopefully everything will work.

Reply:
I resolved the merge conflict and now the build passes.  🚀 

Merging!

Thanks @zanecodes!  Great PR.